pkg/operator/configobservation/apiserver/observer_apiserver.go

```diff
@@ -0,0 +1,199 @@
+		Name:      servingCertSecretName,
+		Namespace: operatorclient.GlobalUserSpecifiedConfigNamespace,
+	}
+	if err = listers.ResourceSyncer().SyncConfigMap(destinationLocation, sourceLocation); err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	return observedConfig, errs
+}
+
+func ObserveNamedCertificates(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	var errs []error
+	namedCertificatesPath := []string{"servingInfo", "namedCertificates"}
+
+	previouslyObservedConfig := map[string]interface{}{}
+	if currentNamedCertificates, _, err := unstructured.NestedSlice(existingConfig, namedCertificatesPath...); len(currentNamedCertificates) > 0 || err != nil {
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if err := unstructured.SetNestedSlice(previouslyObservedConfig, currentNamedCertificates, namedCertificatesPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	observedConfig := map[string]interface{}{}
+	apiServer, err := listers.APIServerLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		glog.Warningf("apiservers.config.openshift.io/cluster: not found")
+		return observedConfig, errs
+	}
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	namedCertificates := apiServer.Spec.ServingCerts.NamedCertificates
+	if len(namedCertificates) > 10 {
+		err := fmt.Errorf("apiservers.config.openshift.io/cluster: spec.servingCerts.namedCertificates cannot have more than 10 entries")
+		recorder.Warningf("ObserveNamedCertificatesFailed", err.Error())
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	secretsMap := map[string]string{}
+	if len(namedCertificates) > 0 {
+		var observedNamedCertificates []interface{}
+		for index, namedCertificate := range namedCertificates {
+			observedNamedCertificate := map[string]interface{}{}
+			if len(namedCertificate.Names) > 0 {
+				if err := unstructured.SetNestedStringSlice(observedNamedCertificate, namedCertificate.Names, "names"); err != nil {
+					return previouslyObservedConfig, append(errs, err)
+				}
+			}
+			sourceSecretName := namedCertificate.ServingCertificate.Name
+			if len(sourceSecretName) == 0 {
+				err := fmt.Errorf("apiservers.config.openshift.io/cluster: spec.servingCerts.namedCertificates[%v].servingCertificate.name not found", index)
+				recorder.Warningf("ObserveNamedCertificatesFailed", err.Error())
+				return previouslyObservedConfig, append(errs, err)
+			}
+			targetSecretName := fmt.Sprintf("user-serving-cert-%03d", index)
+			secretsMap[targetSecretName] = sourceSecretName
+			certFile := fmt.Sprintf("/etc/kubernetes/static-pod-resources/secrets/%v/tls.crt", targetSecretName)
+			if err := unstructured.SetNestedField(observedNamedCertificate, certFile, "certFile"); err != nil {
+				return previouslyObservedConfig, append(errs, err)
```

Comment:
yeah, you need defer funcs for syncs

Reply:
Resources now sync in a `defer func()`, unless `skipResourceSync := true` to indicate that you want to keep the previous resources.